pandas/core/algorithms.py

```diff
@@ -267,22 +267,84 @@ def match(to_match, values, na_sentinel=-1):
     return result
 
 
-def unique1d(values):
+def unique(values):
     """
-    Hash table-based unique
+    Hash table-based unique. uniques are returned in order
+    of appearance. This does NOT sort.
+
+    Parameters
+    ----------
+    values : 1d array-like
+
+    Returns
+    -------
+    unique values.
+    - If the input is a Categorical dtype, the return is a Categorical
+    - If the input is an Index, the return is an Index
+    - If the input is a Series/ndarray, the return will be an ndarray
+
+    Examples
+    --------
+    pd.unique(pd.Series([2, 1, 3, 3]))
+    array([2, 1, 3])
+
+    >>> pd.unique(pd.Series([2] + [1] * 5))
+    array([2, 1])
+
+    >>> pd.unique(Series([pd.Timestamp('20160101'),
+    ...                   pd.Timestamp('20160101')]))
+    array(['2016-01-01T00:00:00.000000000'], dtype='datetime64[ns]')
+
+    >>> pd.unique(pd.Series([pd.Timestamp('20160101', tz='US/Eastern'),
+    ...                      pd.Timestamp('20160101', tz='US/Eastern')]))
+    array([Timestamp('2016-01-01 00:00:00-0500', tz='US/Eastern')],
+          dtype=object)
+
+    >>> pd.unique(pd.Index([pd.Timestamp('20160101', tz='US/Eastern'),
+    ...                     pd.Timestamp('20160101', tz='US/Eastern')]))
+    DatetimeIndex(['2016-01-01 00:00:00-05:00'],
+    ...           dtype='datetime64[ns, US/Eastern]', freq=None)
+
+    >>> pd.unique(list('aabc'))
+    array(['a', 'b', 'c'], dtype=object)
+
+    >>> pd.unique(Series(pd.Categorical(list('aabc'))))
+    0    a
+    1    b
+    2    c
+    dtype: category
+    Categories (3, object): [a, b, c]
+
     """
+
     values = _ensure_arraylike(values)
+
+    # categorical is a fast-path
+    if is_categorical_dtype(values):
+
+        if isinstance(values, ABCSeries):
+            from pandas import Series
+            return Series(values.values.unique(), name=values.name)
```

Comment:
As commented above, I think we shouldn't do the Series wrapping here 

That would also be consistent with `.unique()`. With this branch:

```
In [13]: pd.unique(pd.Series(pd.Categorical(list('aabc'))))
Out[13]: 
0    a
1    b
2    c
dtype: category
Categories (3, object): [a, b, c]

In [14]: pd.Series(pd.Categorical(list('aabc'))).unique()
Out[14]: 
[a, b, c]
Categories (3, object): [a, b, c]
```